components/tidb_query/src/codec/row/v2/row_slice.rs

```diff
@@ -0,0 +1,301 @@
+                null_ids: read_ints_le(&mut data, null_cnt)?,
+                offsets: read_ints_le(&mut data, non_null_cnt)?,
+                values: data,
+            }
+        };
+        Ok(row)
+    }
+
+    /// Search `id` in non-null ids
+    ///
+    /// Returns the `start` position and `offset` in `values` field if found, otherwise returns `None`
+    ///
+    /// # Errors
+    ///
+    /// If the id is found with no offset(It will only happen when the row data is broken),
+    /// `Error::ColumnOffset` will be returned.
+    fn search_in_non_null_ids(&self, id: i64) -> Result<Option<(usize, usize)>> {
+        if !self.id_valid(id) {
+            return Ok(None);
+        }
+        match self {
+            RowSlice::Big {
+                non_null_ids,
+                offsets,
+                ..
+            } => {
+                if let Ok(idx) = non_null_ids.binary_search(&(id as u32)) {
+                    let offset = offsets.get(idx).ok_or(Error::ColumnOffset(idx))?;
+                    let start = if idx > 0 {
+                        offsets[idx - 1] as usize
+                    } else {
+                        0usize
+                    };
+                    return Ok(Some((start, (*offset as usize))));
+                }
+            }
+            RowSlice::Small {
+                non_null_ids,
+                offsets,
+                ..
+            } => {
+                if let Ok(idx) = non_null_ids.binary_search(&(id as u8)) {
+                    let offset = offsets.get(idx).ok_or(Error::ColumnOffset(idx))?;
+                    let start = if idx > 0 {
+                        offsets[idx - 1] as usize
+                    } else {
+                        0usize
+                    };
+                    return Ok(Some((start, (*offset as usize))));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    /// Search `id` in null ids
+    ///
+    /// Returns true if found
+    fn search_in_null_ids(&self, id: i64) -> bool {
+        match self {
+            RowSlice::Big { null_ids, .. } => null_ids.binary_search(&(id as u32)).is_ok(),
+            RowSlice::Small { null_ids, .. } => null_ids.binary_search(&(id as u8)).is_ok(),
+        }
+    }
+
+    fn id_valid(&self, id: i64) -> bool {
+        let upper: i64 = if self.is_big() {
+            i64::from(u32::max_value())
+        } else {
+            i64::from(u8::max_value())
+        };
+        id > 0 && id <= upper
+    }
+
+    fn is_big(&self) -> bool {
+        match self {
+            RowSlice::Big { .. } => true,
+            RowSlice::Small { .. } => false,
+        }
+    }
+}
+
+/// Decodes `len` number of ints from `buf` in little endian
+///
+/// Note:
+/// This method is only implemented on little endianness currently, since x86 use little endianness.
+#[cfg(target_endian = "little")]
+#[inline]
+fn read_ints_le<'a, T>(buf: &mut &'a [u8], len: usize) -> Result<&'a [T]> {
```

Comment:
Maybe better to keep restricting `T` within a numeric type (like `PrimInt`), to minimize the possibility of making mistakes. Otherwise this function is not safe.